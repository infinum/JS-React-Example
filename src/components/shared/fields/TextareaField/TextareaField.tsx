import { FormControl, FormErrorMessage, FormLabel, Textarea, TextareaProps, forwardRef } from '@chakra-ui/react';
import { ErrorMessage } from '@hookform/error-message';
import { FieldErrors } from 'react-hook-form';

interface ITextareaFieldProps extends TextareaProps {
	errors: FieldErrors;
	name: string;
	label: string;
}

export const TextareaField = forwardRef<ITextareaFieldProps, 'textarea'>(function InputField(
	{ errors, name, label, ...rest },
	ref
) {
	return (
		<FormControl my={4} id={name} isInvalid={Boolean(errors[name])}>
			<FormLabel>{label}</FormLabel>
			<Textarea ref={ref} errorBorderColor="red" name={name} {...rest} />
			<FormErrorMessage color="red">
				<ErrorMessage errors={errors} name={name} />
			</FormErrorMessage>
		</FormControl>
	);
});
