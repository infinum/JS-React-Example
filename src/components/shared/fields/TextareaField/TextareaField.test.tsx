import { TextareaField } from '@/components/shared/fields/TextareaField/TextareaField';
import { axe } from 'jest-axe';
import { render, screen } from 'test-utils';

const testLabel = 'Test Label';
const testName = 'Test Name';
const testErrorMessage = 'Test Error Message';

const TestTextareaField = ({ label = testLabel, name = testName, errors = {} }) => {
	return <TextareaField label={label} name={name} errors={errors} />;
};

describe('TextareaField', () => {
	it('should matches snapshot', () => {
		const { asFragment } = render(<TestTextareaField />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should be accessible', async () => {
		const { container } = render(<TestTextareaField />);

		const results = await axe(container);

		expect(results).toHaveNoViolations();
	});

	it('should be accessible when invalid', async () => {
		const errors = { [testName]: { message: testErrorMessage } };

		const { container } = render(<TestTextareaField errors={errors} />);

		const results = await axe(container);

		expect(results).toHaveNoViolations();
	});

	it('should display error message when invalid', () => {
		const testErrorMessage = 'Test Error Message';
		const errors = { [testName]: { message: testErrorMessage } };

		render(<TestTextareaField errors={errors} />);

		expect(screen.getByText(testErrorMessage)).toBeInTheDocument();
	});

	it('should render correct label', () => {
		render(<TestTextareaField />);

		expect(screen.getByText(testLabel)).toBeInTheDocument();
	});

	it('should render correct name', () => {
		render(<TestTextareaField />);

		const textareaField = screen.getByLabelText(testLabel);

		expect(textareaField).toHaveAttribute('name', testName);
	});
});
