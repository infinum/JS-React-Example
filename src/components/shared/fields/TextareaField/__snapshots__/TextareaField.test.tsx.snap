// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaField should matches snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 100%;
  position: relative;
  margin-top: var(--chakra-space-4);
  margin-bottom: var(--chakra-space-4);
}

.emotion-1 {
  display: block;
  text-align: start;
  font-size: var(--chakra-fontSizes-md);
  -webkit-margin-end: var(--chakra-space-3);
  margin-inline-end: var(--chakra-space-3);
  margin-bottom: var(--chakra-space-2);
  font-weight: var(--chakra-fontWeights-medium);
  transition-property: var(--chakra-transition-property-common);
  transition-duration: var(--chakra-transition-duration-normal);
  opacity: 1;
}

.emotion-1:disabled,
.emotion-1[disabled],
.emotion-1[aria-disabled=true],
.emotion-1[data-disabled] {
  opacity: 0.4;
}

.emotion-2 {
  width: 100%;
  min-width: 0px;
  outline: 2px solid transparent;
  outline-offset: 2px;
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  transition-property: var(--chakra-transition-property-common);
  transition-duration: var(--chakra-transition-duration-normal);
  padding-top: var(--chakra-space-2);
  padding-bottom: var(--chakra-space-2);
  min-height: var(--chakra-sizes-20);
  line-height: var(--chakra-lineHeights-short);
  vertical-align: top;
  font-size: var(--chakra-fontSizes-md);
  -webkit-padding-start: var(--chakra-space-4);
  padding-inline-start: var(--chakra-space-4);
  -webkit-padding-end: var(--chakra-space-4);
  padding-inline-end: var(--chakra-space-4);
  height: var(--chakra-sizes-10);
  border-radius: var(--chakra-radii-md);
  border: 1px solid;
  border-color: inherit;
  background: inherit;
}

.emotion-2:disabled,
.emotion-2[disabled],
.emotion-2[aria-disabled=true],
.emotion-2[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-2:hover,
.emotion-2[data-hover] {
  border-color: var(--chakra-colors-gray-300);
}

.emotion-2[aria-readonly=true],
.emotion-2[readonly],
.emotion-2[data-readonly] {
  box-shadow: var(--chakra-shadows-none)!important;
  -webkit-user-select: all;
  -moz-user-select: all;
  -ms-user-select: all;
  user-select: all;
}

.emotion-2[aria-invalid=true],
.emotion-2[data-invalid] {
  border-color: #000000;
  box-shadow: 0 0 0 1px #000000;
}

.emotion-2:focus-visible,
.emotion-2[data-focus-visible] {
  z-index: 1;
  border-color: #3182ce;
  box-shadow: 0 0 0 1px #3182ce;
}

<div
    class="chakra-form-control emotion-0"
    role="group"
  >
    <label
      class="chakra-form__label emotion-1"
      for="Test Name"
      id="Test Name-label"
    >
      Test Label
    </label>
    <textarea
      class="chakra-textarea emotion-2"
      id="Test Name"
      name="Test Name"
    />
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
